schemas/advanced-layouts.md

```diff
@@ -0,0 +1,120 @@
+
+Example uses of advanced layouts include:
+
+* Very large byte arrays that may cross many block boundaries but still present a standard `bytes` kind interface.
+* Maps that can be sharded across many blocks allowing efficient access via standard `map` kind interfaces while allowing extremely large storage.
+* Encrypted block mechanisms that are able to encode and decode data at the data model layer (perhaps in `byte` or `string` form) while presenting as unencrypted kinds when accessed through the schema layer.
+* Novel kind representation formats, such as high-precision or large numbers or alternative string encoding formats. Although size constraints at the language layer for number kinds may represent challenges for large or high-precision number formats.
+
+Advanced layouts necessarily involve a form of logic to perform their data transformations. Many use-cases even necessitate interaction with block loading and storing mechanisms, such as traversal of a large `bytes` array stored across a chain of blocks. There is currently no formal requirement for where this logic resides, how it is loaded, or how it is associated with an IPLD Schema definition. This topic will continue to evolve. There is potential in the future for such logic to be embedded _in_ IPLD blocks themselves via WebAssembly, such that an advanced layout is represented both by the data and the logic required to read it. See [#130](https://github.com/ipld/specs/issues/130) for an early exploration of some of these themes.
+
+## Basic schema definition and use
+
+At its most basic, no properties are mandated for advanced layout. One may be defined simply with the keyword `advanced` in a similar manner to `type`:
+
+```ipldsch
+advanced ROT13
+```
+
+It may then be used as a `representation` for a `type`, from which we can infer the `kind` that we expect.
+
+```ipldsch
+type MyString string representation ROT13
+```
+
+Similarly, an advanced layout implementing a sharded `map` may be defined and used as:
+
+```ipldsch
+advanced ShardedMap
+
+type MyMap { String : Link } representation ShardedMap
+```
+
+From this usage, we may infer that `ShardedMap` can (1) present a familiar `map` kind interface and (2) store `link`s as values, referenced by standard data model `string`s. Other operating modes for `ShardedMap` may also be possible (it may be able to store other value kinds, or it may even be able to act as a `bytes` kind in spite of its name!).
+
+## Root node type definitions
+
+Advanced layouts are designed to abstract data that exists at the data model layer. As such, they may also dictate what they expect from the data that exists at the node their _root_ resides at.
```

Comment:
Is this... a direction we want to go at all?

I think this basically means the schema for the innards of the advanced layout would be "vendored in" (so to speak) the schema of the application using it.

This would be possible, but I don't know if it strikes me as useful.  The functionally relevant relationships to those structures are all inside the advanced layout logic.  So I expect when writing advanced layout logic, we'll often find it useful to have a schema for its insides.  But having that internal schema be replicated/vendored in an application schema that's just consuming the advanced layout is redundant and/or at the wrong level for those purposes.

Having it in the application/consuming schema means some forms of validation could be done without having the advanced layout code.  But I'm not at sure that's useful.  I think mostly not.  I can't really imagine an application that's going to be able to make useful partial progress or degraded but-still-useful functionality out of that.

(We could still execute the idea of "send a request for data that treats the advanced layout as transparent and reaches in and only pulls out a subset of its content"... but I'd expect to do that by sending a different schema that uses those snippets of advanced-layout internals.  Different-schemas-on-same-data is sufficient to do that and seems clear.)

Reply:
An idea that's tempting to toy with is having a link to a separate schema that describes the internals of the ADL.

But, shucks, I think that's going to be a lunge we don't want to make.  That would implies cross-schema linking, and that implies either naively pasting a CID hash -- which is a UX that I don't want to put on people -- or opening the door to requiring some sort a naming system that we... are not going to want to expand our scope to include just for this purpose.

So I think having a separate schema, and handling it in the same mechanism as we use to handle the rendevouz problem for finding the ADL code at all is probably the best we can do.

Reply:
> I think this basically means the schema for the innards of the advanced layout would be "vendored in" (so to speak) the schema of the application using it.

Not necessarily. A single schema may contain the types that are required by the layout. Once complete, I would expect that the `unixfsv2` schema contains the layout schema necessary for its advanced types *except for HAMT* which will be loaded separately.

I’d like to see this be as flexible as possible. *If* the schema defines the layout schemas for an advanced layout, great. If it doesn’t and instead just has a string identifier for it, great.

Reply:
> has a string identifier for it

back to the identification problem - what is this identifier? earlier incarnations of this had:

```
advanced HashMap {
  identifier "IPLD/HashMap/0"
}
```

but then we got all tangled and it resulted in https://github.com/ipld/specs/issues/130 (which continues in this thread).

@warpfork wants to punt this whole identify-code-with-schema-definition thing down the road, and I agree that it would be good to postpone that as much as possible and hardwire things in the meantime or add mapping mechanisms to what we build: `compile-thing foo.ipldsch --map HashMap:/path/to/hashmap/code` or whatever does that mapping of the name that appears in the schema (`HashMap`) and the logic that runs it. So for now, the only signal we have is the term that comes after `advanced`.

Reply:
honestly, the name you give the advanced layout is *already* an identifier as far as libraries that parse the schema and generate a library from it are concerned. i wouldn’t put the `identifier` keyword in the critical path right now, that’s much more helpful down the line when we have some way to resolve globally unique names to implementations.